src/Microsoft.AspNet.Razor.Runtime.Precompilation/CodeAnalysisSymbolBasedTypeInfo.cs

```diff
@@ -0,0 +1,327 @@
+        /// </summary>
+        public static readonly System.Reflection.TypeInfo OpenGenericDictionaryTypeInfo =
+            typeof(IDictionary<,>).GetTypeInfo();
+        private readonly CodeAnalysisSymbolLookupCache _symbolLookup;
+        private readonly ITypeSymbol _type;
+        private readonly ITypeSymbol _underlyingType;
+        private string _fullName;
+        private List<IPropertyInfo> _properties;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="CodeAnalysisSymbolBasedTypeInfo"/>.
+        /// </summary>
+        /// <param name="propertySymbol">The <see cref="IPropertySymbol"/>.</param>
+        /// <param name="symbolLookup">The <see cref="CodeAnalysisSymbolLookupCache"/>.</param>
+        public CodeAnalysisSymbolBasedTypeInfo(
+            [NotNull] ITypeSymbol type,
+            [NotNull] CodeAnalysisSymbolLookupCache symbolLookup)
+        {
+            _symbolLookup = symbolLookup;
+            _type = type;
+            _underlyingType = UnwrapArrayType(type);
+        }
+
+        /// <inheritdoc />
+        public string FullName
+        {
+            get
+            {
+                if (_fullName == null)
+                {
+                    _fullName = GetFullName(_type);
+                }
+
+                return _fullName;
+            }
+        }
+
+        /// <summary>
+        /// The <see cref="ITypeSymbol"/> instance.
+        /// </summary>
+        public ITypeSymbol TypeSymbol => _type;
+
+        /// <inheritdoc />
+        public bool IsAbstract => _type.IsAbstract;
+
+        /// <inheritdoc />
+        public bool IsGenericType
+        {
+            get
+            {
+                return _type.Kind == SymbolKind.NamedType &&
+                    ((INamedTypeSymbol)_type).IsGenericType;
+            }
+        }
+
+        /// <inheritdoc />
+        public bool IsNested => _underlyingType.ContainingType != null;
+
+        /// <inheritdoc />
+        public bool IsPublic
+        {
+            get
+            {
+                return _type.DeclaredAccessibility == Accessibility.Public ||
+                    _type.TypeKind == TypeKind.Array;
+            }
+        }
+
+        /// <inheritdoc />
+        public string Name
+        {
+            get
+            {
+                if (_type.TypeKind == TypeKind.Array)
+                {
+                    return _underlyingType.MetadataName + "[]";
+                }
+
+                return _type.MetadataName;
+            }
+        }
+
+        /// <inheritdoc />
+        public IEnumerable<IPropertyInfo> Properties
+        {
+            get
+            {
+                if (_properties == null)
+                {
+                    _properties = GetProperties(_type, _symbolLookup);
+                }
+
+                return _properties;
+            }
+        }
+
+        /// <inheritdoc />
+        public bool IsTagHelper
```

Comment:
@NTaylorMullen something is messed up with this closed PR. Doesn't show the `IsTagHelper` use in `TagHelperTypeResolver.IsTagHelper()`. Makes sense for this to be here because it's part of the type metadata required.